### PR TITLE
[Relay][MergeComposite] Support TupleGetItem in body of pattern

### PR DESCRIPTION
**Pull request by trevor-m**
This PR adds support to MergeComposite to support TupleGetItemNode in the body of the pattern only. It does not work when the root node is a TupleGetItemNode. We also assume that `tuple` field of TupleGetItem is a Call node and not a Tuple node.

This covers common use cases such as batch norm and get_valid_counts + non_max_suppression.

RFC: https://discuss.tvm.ai/t/rfc-external-codegen-extending-mergecomposite-to-pattern-match-other-node-types/5958/2

@mbaret @comaniac

### CODE REVIEW

File: src/relay/transforms/merge_composite.cc [hunk 1]
```diff
@@ -66,6 +66,31 @@ class MergeCompositeWrapper : public ExprMutator {
...
+    if (pattern->tuple->IsInstance<CallNode>() &&
+        root_node->tuple->IsInstance<CallNode>()) {
+      Expr new_arg;
+      if (call_map->find(pattern->tuple) != call_map->end()) {
+        new_arg = (*call_map)[pattern->tuple];
+      } else {
+        new_arg = ExtractPattern(Downcast<Call>(pattern->tuple),
+                                 Downcast<Call>(root_node->tuple),
+                                 var_map, call_map);
+        call_map->Set(pattern->tuple, new_arg);
```
Comment (mbaret) [inline src/relay/transforms/merge_composite.cc]: I think call_map should now be renamed to expr_map so this is a bit easier to follow.

File: src/relay/transforms/merge_composite.cc [hunk 2]
```diff
@@ -66,6 +66,31 @@ class MergeCompositeWrapper : public ExprMutator {
...
+    if (pattern->index != root_node->index) {
+      return Expr();
+    }
+    if (pattern->tuple->IsInstance<CallNode>() &&
+        root_node->tuple->IsInstance<CallNode>()) {
+      Expr new_arg;
+      if (call_map->find(pattern->tuple) != call_map->end()) {
+        new_arg = (*call_map)[pattern->tuple];
+      } else {
+        new_arg = ExtractPattern(Downcast<Call>(pattern->tuple),
```
Comment (mbaret) [inline src/relay/transforms/merge_composite.cc]: I'm not sure I understand why there's a downcast to a Call here, shouldn't it be a Tuple?

Comment (trevor-m) [reply src/relay/transforms/merge_composite.cc]: Thanks for reviewing!
`pattern->tuple` can be a tuple or it can be any node with a TupleType output. In this call it is a Call node (batch_norm) which has multiple outputs so the output is a TupleType.

File: src/relay/transforms/merge_composite.cc (hunk 1 above)
Comment (trevor-m) [reply src/relay/transforms/merge_composite.cc]: Call map is still only used for call nodes with this PR.

File: src/relay/transforms/merge_composite.cc (hunk 2 above)
Comment (mbaret) [reply src/relay/transforms/merge_composite.cc]: Ah, I get it now, thanks.

File: src/relay/transforms/merge_composite.cc (hunk 1 above)
Comment (mbaret) [reply src/relay/transforms/merge_composite.cc]: Alright, we'll leave it for now then (I'll probably change it in a later PR).

**Comment by zhiics** [comment]
@trevor-m Can you retrigger the CI?

**Comment by zhiics** [comment]
Thanks @trevor-m
